[vicelyav] iP

vicelyav opened the pull request:
### Tira the Cafe Cat :smiley_cat:
Tira is a **CAT** owned by [vicelyav](https://github.com/vicelyav/ip)
Tira is based on a cat in my neighbourhood, whose behaviour can be summarised as:

> She is very fat - My Mom

This chatbot will help you:
* Add, delete, and modify your tasks
* Remember deadline and event timings
* Be a good cat friend

All you need to do is:
1. Follow the recommended task and date formatting
2. Remember to enter the correct command and update the task list
3. Say "bye" to end the chatbot to record your tasks

A short instruction on how to do this:
- [ ] Initialise Tira 
- [ ] Type your command
- [ ] To see your tasks, type list
- [ ] To end the chatbot, type "bye"

If you are a Java programmer, you can practice Java too. Here's a short snippet of the `main` class: 
``` ruby
public class Tira {
    private static final String Directory = "./data";
    private static final String FileName = "./data/Tira.txt";
    private static DateTimeFormatter IN_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd");
    private Ui ui;
    private TaskList tasks;
    private Storage storage;
    private Parser parser;

    /**
     * Initialises the Tira class with the input filepath.
     * Loads task list from the file with the matching filename.
     *
     * @param filePath Filepath of the file to be updated.
     */

    public Tira(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            tasks = new TaskList(storage.load());
        } catch (TiraException e) {
            ui.showLoadingError();
            tasks = new TaskList();
        }
        parser = new Parser(tasks);

    } 
```

Tira's code is also designed to mimic a cat's behaviour, such as:
`     public void bye() {
        System.out.println("Bye. Come back with treats, MIAO!");
    } `

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Change Duke class name

Duke class name changed to Tira
String logo description changed

Commit:
Rename Duke java to Tira Java

Commit:
Add Echo Feature

Echo feature uses printWriter for efficiency.
Loops as long as no "bye" command is written

Commit:
Add Level 2. Add, List

Use ArrayList to keep track of the commands.

Commit:
Level-3 Mark as Done

Add Task class to represent tasks added. Bot messages changed to fit bot personality

Commit:
Add Level 4 Features to Tira

Created Deadline, Event, ToDo classes inheriting from Task class.
More abstraction needed in the main Tira class to make the code more readable
Created an addTask method to streamline the task adding process

Commit:
Add Automatic Testing

Modified the runtest.sh file to contain the working directory.
Tested the runtest.sh file.  Added a .trim to eliminate white space around some words in Task.

Commit:
Add Level-5 Error Handling

Handle Error of:
1. Not having the required number of words for the command.
2. The command not in the command list

Commit:
Add more test cases to Automated testing

Tests for addition of Event, Deadline, and Todo to the List. Tests for marking and unmarking tasks, printing taskList

Commit:
Add Level-6 Delete and Revise Test Cases

Add Delete Task feature to Tira (main class). Add and revised several test cases, remove unrequired white spaces from printed strings.

Commit:
Add methods to streamline main() code

Move all the if-else blocks to different methods so that the main() code looks neat. It will potentially allow for future modifications to be easier.

Commit:
Add File not present handling

Checks if there is a file with a specific directory in the project root.

Commit:
Revert "Add methods to streamline main() code"

This reverts commit 1a22a6de37fa684550e77d5c5b799f434b8e61b0.

Commit:
Separate the if-else blocks into methods

Shifted the if-else blocks to individual static methods so that it is easier to read and modify in the future

Commit:
Add Level - 7 Feature to Save taskList to a file

Use File, FileReader, and FileWriter

Commit:
Merge branch 'branch-Level-7'

Commit:
Add Tira.txt to test for the Save function

Commit:
Add Level-8 Dates and Times function

Modified Deadline and Event task to store LocalDate instead of String for the dates and times. Modified addDeadline() and addEvent() method in Tira to format the correct dates and throw exception if the incorrect format is used.

Commit:
Add skeleton for A-OOP

Created skeleton for the level 8 classes : Storage, Parser, Ui, and TaskList

Commit:
Modify TaskList and Tira to better handle tasks

Rectified addEvent, which previously didn't have a startDate. This clears the previous bug of addEvent not functioning properly when the start and end dates are provided as inputs.

Commit:
Rectify addTask  to handle different commands more efficiently

AddTask can now call markTask, unmarkTask, addToDo, addDeadline, addEvent, and delete using a single taskType string. Each command will print out a string indicating that the command has been completed.

Commit:
Add a Comment in Deadline to test for branch merging

Commit:
Revert "Add a Comment in Deadline to test for branch merging"

This reverts commit 95ebe89da6d50ad7ac8122c40fea7935bdd7cdf5.

Commit:
Delete unused DateTimeFormatter

Commit:
Merge branch 'branch-Level-8'

Commit:
Delete unused DateTimeFormatter

Commit:
Improve Storage class

Add and refined save(), load(), convertToTaskList(), convertTaskToString() method in Storage. Modified Tira class to accommodate for Storage class.

Commit:
Add Ui class

Added show methods for taskList and integrated Ui into the Tira main class.

Fixed a minor bug in Storage class where the formatting of String on txt file was inconsistent.

Commit:
Add Parser class

Created Parser class to handle commands. Modified Tira main class to integrate Parser.

Commit:
Merge branch 'A-MoreOOP' with main branch

Resolved conflict with Tira.java

Commit:
Organise files into packages

Move all the java files into main/java/tira folder.

Commit:
Merge branch 'A-Packages'

Commit:
Merge commit '68c58c12c1d631736a4b25500248a2956171496f'

Commit:
Add gradle support

Commit:
Merge branch 'master' of https://github.com/vicelyav/ip

Commit:
Add partial Javadoc as a test

Commit:
Set up gradle properly

Modified gradle-wrapper properties as per github forum to successfully set up gradle.

Commit:
Add JUnit testing

Modified gradle to accommodate JUnit testing. Created TaskListTest and UiTest classes and ran the testing code.

Commit:
Merge branch 'A-JUnit'

Commit:
Modify build.gradle to allow for jar file creation

Added required plugins and changed the mainClassName.

Commit:
Merge branch 'A-Jar'

Commit:
Merge branch 'add-gradle-support'

# Conflicts:
#	build.gradle
#	gradle/wrapper/gradle-wrapper.properties

Tanziyanghenry commented on src/main/java/tira/Parser.java:
```diff
@@ -0,0 +1,54 @@
+        String [] splitCommand = command.split(" ");
+        String taskType = splitCommand[0];
+        try {
+            switch (taskType) {
```
Shouldn't there be no indentations for `case` clauses?

hj235 replied:
Agreed! Take care of indentations for switch case blocks, intellij IDEA automatically indents the case statements, however this violates the coding style followed by CS2103.

Tanziyanghenry commented on src/main/java/tira/Storage.java:
```diff
@@ -0,0 +1,152 @@
+        return output;
+    }
+
+    public ArrayList<Task> convertToTaskList(ArrayList<String> taskStringList) {
```
Maybe a javadoc comment here would be nice?

hj235 replied:
Agreed! I feel it would provide much clarity explaining what the purpose of the method, as well as what the input `taskStringList` represents.

Tanziyanghenry commented on src/main/java/tira/Storage.java:
```diff
@@ -0,0 +1,152 @@
+
+    //@@hansneddyanto-reused
+    // Hans is my friend and he has allowed me to refer to parts of his code.
+    public String convertTaskToString(Task task) {
```
Maybe a javadoc comment here would be nice?

Tanziyanghenry commented on src/main/java/tira/Storage.java:
```diff
@@ -0,0 +1,152 @@
+import java.time.format.DateTimeFormatter;
+import java.io.BufferedReader;
+
+public class Storage {
```
Maybe a javadoc comment here would be nice?

Tanziyanghenry commented on src/main/java/tira/Storage.java:
```diff
@@ -0,0 +1,152 @@
+            boolean isDone = splitString[1].equals("1");
+            String description = splitString[2];
+
+            switch (taskType) {
```
Shouldn't there be no indentations for `case` clauses?

Tanziyanghenry commented on src/main/java/tira/Tira.java:
```diff
@@ -0,0 +1,82 @@
+
+import tira.task.TaskList;
+
+import java.util.Scanner;
```
Maybe making the imports in order will be better?

Tanziyanghenry commented on src/main/java/tira/Ui.java:
```diff
@@ -0,0 +1,74 @@
+    public Ui() {
+        this.scanner = new Scanner(System.in);
+    }
+    public String read() {
```
Maybe a javadoc comment here would be nice?

Tanziyanghenry commented on src/main/java/tira/task/Task.java:
```diff
@@ -0,0 +1,34 @@
+        return description;
+    }
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
I like how you comment your code for easier understanding

Tanziyanghenry commented on src/main/java/tira/Parser.java:
```diff
@@ -0,0 +1,54 @@
+Parser class:
+1. Splits the initial command and then calls the appropriate function
+if command is list, calls the getTaskList method from TaskList
+if command is delete, cal
```
is the comment incomplete?

Tanziyanghenry commented on src/main/java/tira/Parser.java:
```diff
@@ -0,0 +1,54 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+/*
```
maybe add a space here after import for better presentation

Tanziyanghenry reviewed (commented):
Your code looks good to me!!

hj235 commented on src/main/java/tira/Storage.java:
```diff
@@ -0,0 +1,152 @@
+        FileReader fileReader = null;
+
+        try {
+            if (!file.exists()) { //create new file if the file doesn't exist
+                file.getParentFile().mkdirs();
+                file.createNewFile();
+                return tasks;
+            } else {// a file exists
+                fileReader = new FileReader(file);
+                BufferedReader bufferedReader = new BufferedReader(fileReader);
+                ArrayList<String> lines = new ArrayList<>();
+                String line = bufferedReader.readLine();
+                while (line != null) {
+                    lines.add(line);
+                    if (bufferedReader.readLine() == null) {
+                        break;
+                    }
+                }
+                return this.convertToTaskList(lines);
+            }
```
I like how you handled the case where the file doesn't exist at the start of this try block! :+1:
Perhaps it might be beneficial to extract the code from within the `else` block, since it is the expected normal flow of the program, I feel that this could improve the readability of the code.
```suggestion
            if (!file.exists()) { //create new file if the file doesn't exist
                file.getParentFile().mkdirs();
                file.createNewFile();
                return tasks;
            }
            
            // a file exists
            fileReader = new FileReader(file);
            BufferedReader bufferedReader = new BufferedReader(fileReader);
            ArrayList<String> lines = new ArrayList<>();
            String line = bufferedReader.readLine();
            while (line != null) {
                lines.add(line);
                if (bufferedReader.readLine() == null) {
                    break;
                }
            }
            return this.convertToTaskList(lines);
```

hj235 commented on src/main/java/tira/Tira.java:
```diff
@@ -0,0 +1,82 @@
+ */
+
+public class Tira {
+    private static final String Directory = "./data";
+    private static final String FileName = "./data/Tira.txt";
```
It might be good to use SCREAMING_SNAKE_CASE for these constant names instead.
```suggestion
    private static final String DIRECTORY = "./data";
    private static final String FILENAME = "./data/Tira.txt";
```

hj235 commented on src/main/java/tira/TiraException.java:
```diff
@@ -0,0 +1,9 @@
+    public TiraException(String message) {
+        super(message);
+    }
+    // will revise the exceptions later
```
It's nice to know that the exception classes will be later revised.
Keep up the good work! :+1:

hj235 commented on src/main/java/tira/Ui.java:
```diff
@@ -0,0 +1,74 @@
+Handles printing of all the errors.
+Used as an input for TaskList, Storage, and Tira Main.
+ */
+public class Ui {
```
Perhaps it would be neater to use line breaks consistently between each segment of the code, for example between all method code blocks, rather than only some of them.

hj235 commented on src/main/java/tira/task/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+        if (splitCommand.length < 2) {
+            throw new TiraException("MRAW?? WHERE IS THE TASK?");
+        }
+        int currNum = Integer.parseInt(splitCommand[1]) - 1;
+        tasks.get(currNum).markStatus();
+        Task currTask = tasks.get(currNum);
```
I'm not too sure I like the naming of `currNum` and `currTask`. Perhaps a more intuitive name could be used here, such as `index` and `currentTask`.

hj235 reviewed (commented):
LGTM! 👍 
Some cosmetic changes requested.

Commit:
Add JavaDoc comments to half of the class

Added JavaDoc comments for all classes except for TaskList, Parser, Ui and Deadline

Commit:
Modify code to follow coding standard

Modified order of imports so that they are easier to read. Edited some spacing and comments.

Commit:
Add Find method

Modified TaskList and Parser to handle find method. Parser checked if the command is "find" and TaskList checked which tasks are matching. Ui in TaskList prints out the matching tasks.

Commit:
Merge branch 'A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/tira/task/TaskList.java

Commit:
Fix bug in addEvent

Edited the case-switch block in Ui if the command is event. Previously it was addDeadline() instead of addEvent

Commit:
Merge branch 'branch-Level-9'

Commit:
Fix bug in TaskList and add Launcher and Main classes for JavaFX

Fix problem where findTask is not found in TaskList class. Added skeleton for launcher and Main class to commit before further changes.

Commit:
Fix build.gradle for Step 1 GUI tutorial

Commit:
Modify Main, DialogBox, and Application according to Step 2 JavaFX tutorial

Added formatting codes in Main, DialogBox so that it looks like a chat box. Also added images for the user and the chatbot.

Commit:
Modify DialogBox, MainWindow, Main classes and add MainWindow and DialogBox fxml

Added the appropriate codes in the DialogBox, MainWindow, Main classes and incorporated fxml to fix the chatbot app appearance. Used SceneBuilder to modify background color of the input box.

Commit:
Merge branch 'branch-Level-10'

Commit:
Add Checkstyle, checkstyle.xml and suppressions.xml

No checks on the coding style and quality.

Coding style and readability is important.

Adding checkstyle plugin.

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add assertions for task number during marking and unmarking tasks

Marking and unmarking tasks don't check for positive task number. This
could lead to indexing error while accessing the tasks from the
ArrayList of tasks.

Let's use assertion to ensure that the task number being marked and
unmarked is positive and non-zero in the taskList class.

Also refactored several import lines.

Commit:
Modify coding style using CheckStyle

There are multiple indentation, organisation, and lack of javadoc
comments in the entire code.

This reduces readability for other users.

Let's use CheckStyle plugin of Gradle to improve the coding style.

This improves the efficiency in rectifying the coding style.

Commit:
Rename and remove redundant variables

Several variables are not descriptive and might confuse readers
regarding their function. Several variables are also not used.

Let's standardise variables by:
1. Having all Ui methods that print messages start with show.
2. Specify how the commands are split based on their regex.

Additional variables are also added to improve readability.

vicelyav committed:
Merge branch 'master' into branch-A-CodeQuality

vicelyav committed:
Merge pull request #1 from vicelyav/branch-A-CodeQuality

Deleted conflicting import statements in UiTest task

vicelyav committed:
Merge branch 'master' into branch-A-Assertions

vicelyav committed:
Merge pull request #2 from vicelyav/branch-A-Assertions

Add assertions for task number during marking and unmarking tasks in TaskList class

Commit:
Add missing variables and arguments to TaskList methods

Argument initially named splitCommand missing from the TaskList methods,
causing indexing errors.

Let's add back the arguments and rename them to commandSplitBySpace
to make it more descriptive to readers.

Commit:
Fix bug where not all lines from text file are read

Only one line of the text file is read and it is read multiple times.
This leads to inaccurate ArrayList of Tasks extracted from the file.
There is no new line character added to the text lines, causing the
task sentences to be combined in a single line in the text file.

Let's remove the duplicate readLine() method of bufferedReader in the
load() method of Storage and add "\n" to the task sentences.

Also fix coding quality issues including indentation, white spaces
and refactoring startDate and endDate to two separate methods in
Storage class.

Commit:
Add Statistics Extension

There are no methods that summarises the statistics of the tasks.
Having statistics will be useful for users to gauge their progress
without looking at each task specifically.

Let's,
* use a Statistics class to store and calculate the statistics of the
tasks
* add showStatistics() in Ui class to display the statistics in String
form

Separate Statistics class will provide abstraction, allowing for easier
modification of statistics method in the future. Having Ui show the
statistics also help abstraction, allowing Ui to handle most printing
tasks to the users.

Also modifies several formatting issues for dates of tasks that led
to indexing and formatting errors.

Commit:
Fix GUI format not showing picture properly

GUI top pane hides the previous messages and profile picture. This
prevents user from seeing the previous messages clearly.

Let's fix this by replacing the fmxl file with the original fmxl file,
then tweak the design using Scene Builder.

Commit:
Fix GUI not reflecting Tira functionality

GUI only mirrors the user's command but does not process them. This
limits the functionality of the chatbot.

Let's
* modify the Ui class to be able to return a String.
* modify the Tira getResponse() method so it returns a String.
* modify the MainWindow to handle bye and other commands.

This will ensure that a String is sent back to the GUI, without
modifying much of the existing implementation.

Commit:
Improve GUI readability

Text box border overlaps with the font, making it difficult to read.

Let's modify the DialogBox fmxl so that the border is less rounded and
has space from the font. Let's also add Tira the Cafe Cat description
to the stage.

Commit:
Add more exception handling for Task classes

There is no exception to handle lack of description and dates for ToDo,
Deadline and Event classes. This causes the tasks to still be recorded
in the TaskList despite the error, which leads to formatting errors
for other methods. Delete method also doesn't handle number input
being smaller than TaskList size.

Let's handle these through these methods:
* check if there's 1 or more word after todo command.
* check if the 2nd word in deadline doesn't start with / which indicates
date.
* check if command after split using slash creates an array of more than
2 words.
* conduct similar check for Event as with Deadline class.
* check if the size of the TaskList is greater than the deleted task
number.

Also refactor several lines using substring to differentiate /from, /to,
and /by using extractAfterWord method in TaskList.

Commit:
Add Exception handling for invalid task number for mark and unmark

The mark and unmark doesn't handle cases where the number in user input
is greater than the task list or when number is not included in the
command.

Let's add Exception for these cases by checking against the TaskList
size.

Commit:
Add product screenshot, minor bug fixes and Checkstyle

There is no product screenshot yet and multiple coding style violation,
the latter reducing the readability of the code.

Let's use Checkstyle to rectify the coding style and add product
screenshot into the docs folder.

Commit:
Fix Statistics bug and Add User Guide

Statistics calculate tasks wrongly and there is no user guide. This
makes the statistics result inaccurate. Users cannot operate the
application freely without the user guide.

Let's refactor the Statistics addition of tasks to a single method
and marked or unmarked tasks to another method to make it easier
to read and modify in the future. We also add user guide based on the
commands that the users can input.

Commit:
Add exit, save and editing guide in the user guide

There is no guide on how to exit or save the data in the current user
guide.

Let's add the commands and formats for exit and instructions on how
to save the data.

Commit:
Add JavaDoc for Statistics class

Statistics don't have JavaDoc that explains the methods and
constructors, which will make users confused.

Let's add JavaDoc to the methods.

Commit:
Update DialogBox to have images with square ratio

Image is not in a square ratio which does not look neat to users.

Let's set display picture width and hight to 100 and keep the ratio in
DialogBox.

Commit:
Replace Ui.png with the updated Ui screenshot

The ui image does not reflect the changes made in DialogBox class.

Let's screenshot a new GUI image and place it in the docs folder.

Commit:
Fix wrong format for exit command in readme

Exit command is "exit" instead of "bye"

Let's replace the "exit" format to "bye"

Commit:
Add Exception handling for incorrect LocalDate format

There is improper exception handling for incorrect LocalDate input by
user in deadline and event. User might not understand where they went
wrong.

Let's throw a TiraException with appropriate message so the exception
is handled more elegantly.

Also remove redundant properties in Tira class.

Commit:
Remove unused imports

There is unused DateTimeFormatter in TaskList and Tira class, which
makes the code less neat.

Let's delete the unused import.

Commit:
Fix typo in user guide

There is a misplaced slash in the Deadline section of user guide.

Let's fix the typo and make the date formatting correct.

Commit:
Add more Javadoc

Certain classes have missing Javadoc return lines, which makes reading
them confusing.

Let's add more Javadoc lines to make the code more readable.
